### ISSUE TITLE
Feature request: Teach SpotBugs what Thread.holdsLock() means

### ISSUE THREAD

**Issue by archiecobbs** [spotbugs/spotbugs#2121]
Originally posted in [this thread](https://mailman.cs.umd.edu/pipermail/findbugs-discuss/2016-August/004305.html) and rejected by FindBugs in [this issue](https://sourceforge.net/p/findbugs/feature-requests/360/) (presumably because FindBugs is lame/dead).

I have some code that looks like this:

```java
public class ClassA {
    @GuardedBy("this")
    private int somefield;
}

public class ClassB {
    private final ClassA classA;
    ...
    void foo() {
        assert Thread.holdsLock(this.classA);
        ... this.classA.somefield ...        // false positive unsynchronized access reported here
    }
```

SpotBugs is complaining of `IS_FIELD_NOT_GUARDED`.

It would be nice if SpotBugs could realize that reaching that point in the code without holding the lock that guards the field is impossible due to the assertion.

So can we teach SpotBugs the meaning of `Thread.holdsLock()`?

**Comment by ThrawnCA** [comment]
Assertions are not always enforced, so SpotBugs isn't necessarily wrong to ignore them.

What code is there to ensure that the thread will, in fact, hold the lock, and that the assertion will pass?

**Comment by archiecobbs** [comment]
> Assertions are not always enforced, so SpotBugs isn't necessarily wrong to ignore them.

Apologies for mixing two different things together.

Whether (a) SpotBugs "believes" assertions and whether (b) it understands `Thread.holdsLock()` are two separate/orthgonal questions.

So forget (a) for now and let's focus on (b).

Consider this example below instead of the previous one:
```java
import javax.annotation.concurrent.GuardedBy;
public class ClassA {

    @GuardedBy("this")
    boolean foobar;

    public synchronized boolean isFoobar() {
        return this.foobar;
    }
    public synchronized void setFoobar(final boolean foobar) {
        this.foobar = foobar;
    }

    public void donkey() {
    }
}
```
```java
public abstract class ClassB {

    final ClassA raft;

    ClassB(ClassA raft) {
        this.raft = raft;
    }

    void handleElectionTimeout() {
        if (Thread.holdsLock(this.raft) && this.raft.foobar)
            this.raft.donkey();
    }
}
```
Running SpotBugs I get this:
```
$ mvn clean verify
...
[INFO] --- spotbugs-maven-plugin:4.7.1.0:check (default) @ bug ---
[INFO] BugInstance size is 1
[INFO] Error size is 0
[INFO] Total bugs: 1
[ERROR] High: foobar not guarded against concurrent access; locked 66% of time [ClassB, ClassA, ClassA] Unsynchronized access at ClassB.java:[line 10]Synchronized access at ClassA.java:[line 8]Synchronized access at ClassA.java:[line 11] IS_FIELD_NOT_GUARDED
[INFO] 


To see bug detail using the Spotbugs GUI, use the following command "mvn spotbugs:gui"

```

Is it possible SpotBugs can be smart enough to detect that all control flows through the access to `this.raft.foobar` must first pass through `Thread.holdsLock(this.raft)`, and therefore the access must be synchronized ?
